custom-targets/vertex-ai/model-deployer/deploy.go

```diff
@@ -0,0 +1,304 @@
+			ResultStatus:   clouddeploy.DeployFailed,
+			FailureMessage: err.Error(),
+		}
+		d.addCommonMetadata(res)
+		fmt.Println("Uploading failed deploy results")
+		rURI, err := d.req.UploadResult(ctx, d.gcsClient, dr)
+		if err != nil {
+			return fmt.Errorf("error uploading failed deploy results: %v", err)
+		}
+		fmt.Printf("Uploaded failed deploy results to %s\n", rURI)
+		return err
+	}
+	d.addCommonMetadata(res)
+
+	fmt.Println("Uploading successful deploy results")
+	rURI, err := d.req.UploadResult(ctx, d.gcsClient, res)
+	if err != nil {
+		return fmt.Errorf("error uploading deploy results: %v", err)
+	}
+	fmt.Printf("Uploaded deploy results to %s\n", rURI)
+	return nil
+
+}
+
+func (d *deployer) deploy(ctx context.Context) (*clouddeploy.DeployResult, error) {
+	localManifest := "manifest.yaml"
+	fmt.Printf("Downloading deploy input manifest from %q.\n", d.req.ManifestGCSPath)
+
+	_, err := d.req.DownloadManifest(ctx, d.gcsClient, localManifest)
+	if err != nil {
+		return nil, fmt.Errorf("unable to download deploy input from %s: %v", d.req.ManifestGCSPath, err)
+	}
+
+	fmt.Printf("Downloaded deploy input manifest.\n")
+
+	manifestData, err := d.applyModel(ctx, localManifest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to deploy model: %v", err)
+	}
+
+	mURI, err := d.req.UploadArtifact(ctx, d.gcsClient, "manifest.yaml", &clouddeploy.GCSUploadContent{Data: manifestData})
+	if err != nil {
+		return nil, fmt.Errorf("error uploading deploy artifact: %v", err)
+	}
+
+	return &clouddeploy.DeployResult{
+		ResultStatus:  clouddeploy.DeploySucceeded,
+		ArtifactFiles: []string{mURI},
+	}, nil
+}
+
+func (d *deployer) addCommonMetadata(rs *clouddeploy.DeployResult) {
+	if rs.Metadata == nil {
+		rs.Metadata = map[string]string{}
+	}
+	rs.Metadata[clouddeploy.CustomTargetSourceMetadataKey] = aiDeployerSampleName
+	rs.Metadata[clouddeploy.CustomTargetSourceSHAMetadataKey] = clouddeploy.GitCommit
+}
+
+func deployModelFromManifest(path string) (*aiplatform.GoogleCloudAiplatformV1DeployModelRequest, error) {
+	data, err := os.ReadFile(path)
+	if err != nil {
+		return nil, fmt.Errorf("error reading manifest file: %v", err)
+	}
+
+	deployModelRequest := &aiplatform.GoogleCloudAiplatformV1DeployModelRequest{}
+
+	if err = yaml.Unmarshal(data, deployModelRequest); err != nil {
+		return nil, fmt.Errorf("unable to parse deploy model deployModelRequest from manifest file: %v", err)
+	}
+
+	return deployModelRequest, nil
+}
+
+func (d *deployer) applyModel(ctx context.Context, localManifest string) ([]byte, error) {
+
+	deployModelRequest, err := deployModelFromManifest(localManifest)
+	if err != nil {
+		return nil, err
+	}
+
+	endpointRegion, err := fetchRegionFromEndpoint(d.params.endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse region from endpoint resource name: %v", err)
+	}
+
+	aiplatformService, err := newService(ctx, endpointRegion)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create service: %v", err)
+	}
```

Comment:
This can be moved to a "deployer" constructor and have it be on the object itself. Same for the other request handler implementations.